File: Month03_FSharp/Workers/Worker.fs
namespace Workers

open System
open System.Collections.Concurrent
open System.Diagnostics
open System.Threading
open System.Threading.Channels
open System.Threading.Tasks
open Workers.Incoming

type Worker(reader: ChannelReader<Work>) =

    // Demo "exactly-once" ledger (process lifetime only).
    static let processed = ConcurrentDictionary<int, byte>()

    member public this.Execute(cancellationToken: CancellationToken) : Task =
        task {
            let mutable work = Unchecked.defaultof<Work>

            while reader.TryRead(&work) do
                Console.WriteLine(
                    $"Working on item {work.Id} | Model={work.ExecutionModel:G} | Behaviour={work.Behaviour:F}"
                )

                do! Worker.Simulate(work, cancellationToken)

                Console.WriteLine($"Done work {work.Id}.")
        }

    static member private Simulate(work: Work, cancellationToken: CancellationToken) : Task =
        task {
            // Exactly-once: skip duplicates (demo-level, in-memory only).
            if
                work.Behaviour.HasFlag(Behaviour.ExactlyOnce)
                && not (processed.TryAdd(work.Id, 0uy))
            then
                Console.WriteLine($"  [ExactlyOnce] Work {work.Id} already processed -> skipping.")
            else
                let attempts = if work.Behaviour.HasFlag(Behaviour.Retryable) then 3 else 1

                for attempt in 1..attempts do
                    try
                        cancellationToken.ThrowIfCancellationRequested()

                        // HighPriority: less "queueing"/overhead before work begins.
                        if work.Behaviour.HasFlag(Behaviour.HighPriority) then
                            Console.WriteLine("  [HighPriority] Fast-lane execution.")

                        // RequiresAffinity: pin to a pretend partition/worker lane.
                        let lane = Math.Abs(work.Id) % 4

                        if work.Behaviour.HasFlag(Behaviour.RequiresAffinity) then
                            Console.WriteLine($"  [RequiresAffinity] Routing to lane {lane}.")

                        do! Worker.SimulateByExecutionModel(work, lane, cancellationToken)

                        Console.WriteLine(
                            if work.Behaviour.HasFlag(Behaviour.Retryable) then
                                $"  [Retryable] Succeeded on attempt {attempt}/{attempts}."
                            else
                                "  Completed."
                        )

                    with
                    | :? OperationCanceledException -> Console.WriteLine("  Work cancelled")
                    | ex ->
                        Console.WriteLine($"  Attempt {attempt}/{attempts} failed: {ex.Message}")

                        if attempt = attempts then
                            Console.WriteLine("  Giving up.")
                        else
                            // Tiny exponential-ish backoff for the demo.
                            let backoffMs = 50 * attempt * attempt

                            do! Task.Delay(backoffMs, cancellationToken)

                            Console.WriteLine($"  Work cancelled{ex.Message}")
        }

    static member private SimulateByExecutionModel(work: Work, lane: int, cancellationToken: CancellationToken) : Task =
        task {
            // Base pacing. Behaviours can tweak it.
            let stepDelayMs =
                match work.Behaviour with
                | Behaviour.LongRunning ->
                    Console.WriteLine("  [LongRunning] Slower steps.")
                    240
                | _ -> 80

            if work.Behaviour.HasFlag(Behaviour.ResourceIntensive) then
                Console.WriteLine("  [ResourceIntensive] Adding CPU work.")

            // Scheduled: pretend we had to wait until a trigger time.
            if work.ExecutionModel = ExecutionModel.Scheduled then
                Console.WriteLine("  [Scheduled] Waiting for trigger...")
                do! Task.Delay(150, cancellationToken)

            match work.ExecutionModel with
            | ExecutionModel.OneOff ->
                do! Worker.Step("OneOff: run once", work.Behaviour, stepDelayMs, cancellationToken)
            | ExecutionModel.EventDriven ->
                do! Worker.Step("EventDriven: handle event payload", work.Behaviour, stepDelayMs, cancellationToken)
            | ExecutionModel.Batch ->
                Console.WriteLine("  [Batch] Processing items...")

                for i in 1..5 do
                    do! Worker.Step($"Batch item {i}/5", work.Behaviour, stepDelayMs, cancellationToken)
            | ExecutionModel.Stream ->
                Console.WriteLine("  [Stream] Polling/consuming stream ticks...")

                for i in 1..4 do
                    do! Worker.Step($"Stream tick {i}/5", work.Behaviour, stepDelayMs + 40, cancellationToken)
            | ExecutionModel.Workflow ->
                Console.WriteLine("  [Workflow] Running steps (DAG-ish)...")

                do! Worker.Step("Step A: validate", work.Behaviour, stepDelayMs, cancellationToken)
                do! Worker.Step("Step B: transform", work.Behaviour, stepDelayMs + 20, cancellationToken)
                do! Worker.Step("Step C: persist", work.Behaviour, stepDelayMs + 40, cancellationToken)
            | ExecutionModel.Actor ->
                Console.WriteLine($"  [Actor] Handling partition key lane={lane} (stateful-ish).")

                do! Worker.Step("Actor turn: load state", work.Behaviour, stepDelayMs, cancellationToken)
                do! Worker.Step("Actor turn: apply work", work.Behaviour, stepDelayMs + 20, cancellationToken)
                do! Worker.Step("Actor turn: save state", work.Behaviour, stepDelayMs + 40, cancellationToken)
            | ExecutionModel.Scheduled ->
                do! Worker.Step("Scheduled: execute job", work.Behaviour, stepDelayMs, cancellationToken)
            | _ -> do! Worker.Step("Unknown model: fallback", work.Behaviour, stepDelayMs, cancellationToken)
        }

    static member private Step
        (label: string, behaviour: Behaviour, delayMs: int, cancellationToken: CancellationToken)
        : Task =
        cancellationToken.ThrowIfCancellationRequested()

        Console.WriteLine($"  -> {label}")

        if behaviour.HasFlag(Behaviour.ResourceIntensive) then
            Worker.CpuBump(25_000)

        Worker.MaybeFailTransiently(behaviour)
        Task.Delay(delayMs, cancellationToken)

    // Random transient failure for Retryable demos.
    static member private MaybeFailTransiently(behaviour: Behaviour) : unit =
        if behaviour.HasFlag(Behaviour.Retryable) then
            ()
        else
            // Fail sometimes (more often when "resource intensive") so retries are visible.
            let odds =
                if behaviour.HasFlag(Behaviour.ResourceIntensive) then
                    4
                else
                    7 // 1/4 or 1/7

            if Random.Shared.Next(odds) = 0 then
                raise (InvalidOperationException("Transient failure (simulated)."))

    // ResourceIntensive: do a tiny CPU spin per step (demo only).
    static member private CpuBump(iterations: int) : unit =
        let sw = Stopwatch.StartNew()

        let x = seq { 0..iterations } |> Seq.fold (fun acc i -> (acc * 31) ^^^ i) 0
        ignore x

        sw.Stop()

File: 00_CSharp/Workers/Worker.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Threading;
using System.Threading.Channels;
using System.Threading.Tasks;
using Workers.Incoming;

namespace Workers;

internal readonly record struct Worker
{
    // Demo "exactly-once" ledger (process lifetime only).
    private static readonly ConcurrentDictionary<int, byte> Processed = new();

    private readonly ChannelReader<Work> _reader;

    public Worker(ChannelReader<Work> reader) => _reader = reader;

    public async Task Execute(CancellationToken cancellationToken)
    {
        while (_reader.TryRead(out var work))
        {
            Console.WriteLine(
                $"Working on item {work.Id} | Model={work.ExecutionModel:G} | Behaviour={work.Behaviour:F}");

            await SimulateAsync(work, cancellationToken);

            Console.WriteLine($"Done work {work.Id}.");
        }
    }

    private static async Task SimulateAsync(Work work, CancellationToken ct)
    {
        // Exactly-once: skip duplicates (demo-level, in-memory only).
        if (work.Behaviour.HasFlag(Behaviour.ExactlyOnce) && !Processed.TryAdd(work.Id, 0))
        {
            Console.WriteLine($"  [ExactlyOnce] Work {work.Id} already processed -> skipping.");
            return;
        }

        var attempts = work.Behaviour.HasFlag(Behaviour.Retryable) ? 3 : 1;

        for (var attempt = 1; attempt <= attempts; attempt++)
        {
            try
            {
                ct.ThrowIfCancellationRequested();

                // HighPriority: less "queueing"/overhead before work begins.
                if (work.Behaviour.HasFlag(Behaviour.HighPriority))
                {
                    Console.WriteLine("  [HighPriority] Fast-lane execution.");
                }

                // RequiresAffinity: pin to a pretend partition/worker lane.
                var lane = Math.Abs(work.Id) % 4;
                if (work.Behaviour.HasFlag(Behaviour.RequiresAffinity))
                {
                    Console.WriteLine($"  [RequiresAffinity] Routing to lane {lane}.");
                }

                await SimulateByExecutionModelAsync(work, lane, ct);

                Console.WriteLine(
                    work.Behaviour.HasFlag(Behaviour.Retryable)
                        ? $"  [Retryable] Succeeded on attempt {attempt}/{attempts}."
                        : "  Completed.");

                return;
            }
            catch (OperationCanceledException)
            {
                Console.WriteLine($"  Work cancelled");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  Attempt {attempt}/{attempts} failed: {ex.Message}");

                if (attempt == attempts)
                {
                    Console.WriteLine("  Giving up.");
                    return;
                }

                // Tiny exponential-ish backoff for the demo.
                var backoffMs = 50 * attempt * attempt;
                await Task.Delay(backoffMs, ct);
            }
        }
    }

    private static async Task SimulateByExecutionModelAsync(Work work, int lane, CancellationToken ct)
    {
        // Base pacing. Behaviours can tweak it.
        var stepDelayMs = 80;

        if (work.Behaviour.HasFlag(Behaviour.LongRunning))
        {
            stepDelayMs += 160;
            Console.WriteLine("  [LongRunning] Slower steps.");
        }

        if (work.Behaviour.HasFlag(Behaviour.ResourceIntensive))
        {
            Console.WriteLine("  [ResourceIntensive] Adding CPU work.");
        }

        // Scheduled: pretend we had to wait until a trigger time.
        if (work.ExecutionModel == ExecutionModel.Scheduled)
        {
            Console.WriteLine("  [Scheduled] Waiting for trigger...");
            await Task.Delay(150, ct);
        }

        switch (work.ExecutionModel)
        {
            case ExecutionModel.OneOff:
                await StepAsync("OneOff: run once", stepDelayMs);
                break;

            case ExecutionModel.EventDriven:
                await StepAsync("EventDriven: handle event payload", stepDelayMs);
                break;

            case ExecutionModel.Batch:
                Console.WriteLine("  [Batch] Processing items...");
                for (var i = 1; i <= 5; i++)
                {
                    await StepAsync($"Batch item {i}/5", stepDelayMs);
                }
                break;

            case ExecutionModel.Stream:
                Console.WriteLine("  [Stream] Polling/consuming stream ticks...");
                for (var tick = 1; tick <= 4; tick++)
                {
                    await StepAsync($"Stream tick {tick}/4", stepDelayMs + 40);
                }
                break;

            case ExecutionModel.Workflow:
                Console.WriteLine("  [Workflow] Running steps (DAG-ish)...");
                await StepAsync("Step A: validate", stepDelayMs);
                await StepAsync("Step B: transform", stepDelayMs + 20);
                await StepAsync("Step C: persist", stepDelayMs + 40);
                break;

            case ExecutionModel.Actor:
                Console.WriteLine($"  [Actor] Handling partition key lane={lane} (stateful-ish).");
                await StepAsync("Actor turn: load state", stepDelayMs);
                await StepAsync("Actor turn: apply work", stepDelayMs + 20);
                await StepAsync("Actor turn: save state", stepDelayMs + 40);
                break;

            case ExecutionModel.Scheduled:
                await StepAsync("Scheduled: execute job", stepDelayMs);
                break;

            default:
                await StepAsync("Unknown model: fallback", stepDelayMs);
                break;
        }
        return;

        // Random transient failure for Retryable demos.
        void MaybeFailTransiently()
        {
            if (!work.Behaviour.HasFlag(Behaviour.Retryable))
            {
                return;
            }

            // Fail sometimes (more often when "resource intensive") so retries are visible.
            var odds = work.Behaviour.HasFlag(Behaviour.ResourceIntensive) ? 4 : 7; // 1/4 or 1/7
            if (Random.Shared.Next(odds) == 0)
            {
                throw new InvalidOperationException("Transient failure (simulated).");
            }
        }

        // ResourceIntensive: do a tiny CPU spin per step (demo only).
        static void CpuBump(int iterations)
        {
            var sw = Stopwatch.StartNew();
            var x  = 0;
            for (var i = 0; i < iterations; i++)
            {
                x = (x * 31) ^ i;
            }
            _ = x;
            sw.Stop();
        }

        Task StepAsync(string label, int delayMs)
        {
            ct.ThrowIfCancellationRequested();

            Console.WriteLine($"  -> {label}");
            if (work.Behaviour.HasFlag(Behaviour.ResourceIntensive))
            {
                CpuBump(25_000);
            }

            MaybeFailTransiently();
            return Task.Delay(delayMs, ct);
        }
    }
}